### ISSUE TITLE
AMDGPUCodeGenPrepare should stop promoting uniform i16 values

### ISSUE THREAD

**Issue by arsenm** [llvm/llvm-project#64591]
With DAG divergence and RegBankSelect, theoretically codegen should now have enough information to handle promotion of uniform 16-bit ops directly. This was originally done as a workaround for having no uniformity info after the IR.

AMDGPUCodeGenPrepare awkwardly uses forward iteration, which just so happens to work around a defect in uniformity info where newly created values are assumed uniform. I currently have a problem that requires iterating in reverse like a normal combiner, and this is blocking the reversal

**Comment by kzhuravl** [comment]
Need to re-assign to Jun.

**Comment by kzhuravl** [comment]
Jun is working on other issues. Pierre, can you take a look at this one?

**Comment by Pierre-vh** [comment]
@arsenm Is it just a matter of removing all i16 promotion logic from AMDGPUCodeGenPrepare (as-if Widen16BitOps was always false) or are there other things to do as well?

**Comment by arsenm** [comment]
> @arsenm Is it just a matter of removing all i16 promotion logic from AMDGPUCodeGenPrepare (as-if Widen16BitOps was always false) or are there other things to do as well?
> 
> Is this about moving all of the i16 promotion logic to DAG/GISel, or is that already implemented?

Yes, I assume other work needs to be done to avoid regressions. I don't think any of the relevant work has been implemented in the codegen side, it's just more tractable with DAG divergence info

**Comment by Pierre-vh** [comment]
@arsenm I got started on this, but I'm having some trouble.

I initially started implementing this as a new `LowerOperation` case, by adding a `Custom` legalization to all of the operations I wanted to look at. This more or less works but I run into issues because sometimes we don't want to promote (when the value is uniform), but the operation is still marked as illegal and:
 - if I return `SDValue()`, the legalizer will try to expand/fallback to something else, making things worse.
 - if I return the original node, an assertion failure happens for some operations as "potential combine loop" is detected, IIRC.

I then tried making it a combine, but it's even worse, it gets into all sorts of combine loops because the combiner really hates `trunc (op ...)` when op can be narrowed.

So I assume the Custom legalization path is the correct tone, but I can't figure out how to tell the legalizer that we intentionally didn't lower an operation?

**Comment by arsenm** [comment]
Presumably you need to make isNarrowingProfitable aware of the specific node (such that we can check the divergence bit), and then do something to visitTRUNCATE to prevent it from narrowing

**Comment by jayfoad** [comment]
@Pierre-vh was there any progress on this? And do you have a feeling for the scope of the work, i.e. are there known regressions if you disable the promotion code, or is this more about hunting for the unknown regressions?

**Comment by Pierre-vh** [comment]
I got stuck trying to implement the promotion logic in GlobalISel. I think it has to be done in the legalizer, but the legalizer needs UniformityInfo to do it. Whenever I tried doing it somewhere else I ran into issues, either combine loops or generating illegal instructions too late.

**Comment by jayfoad** [comment]
Are you saying that if you turn off the promotion in AMDGPUCodeGenPrepare, then GlobalISel does not even generate functionally correct code?

**Comment by Pierre-vh** [comment]
> then GlobalISel does not even generate functionally correct code?

Why would the code be incorrect ? Do we depend on i16->i32 promotion in CGP for correctness?

**Comment by Pierre-vh** [comment]
Ah, sorry. I believe the generated code is still correct, but slower. This widening was done in CGP because we initially had no divergence info in codegen, but that's no longer the case. Doing the widening during codegen allows us to find more opportunities to do it.

I think if we turn off the widening in CGP now, we might see some regressions in the DAG, and a lot more regressions in GlobalISel.

Is there a need to make the switch as soon as possible ?

**Comment by jayfoad** [comment]
> Is there a need to make the switch as soon as possible ?

It's on my radar because I saw some f16 arithmetic micro-benchmarks which would benefit from turning off the promotion in CGP.

**Comment by Pierre-vh** [comment]
I was curious and looked at disabling the option right now, and there are a few DAGISel cases with big gains, e.g. `cttz_zero_undef.ll`'s has a function where 5-6 instructions become just one `v_ffbl_b32_e32`. Some other cases have regressions though, I'll check if we can add new combines to fix them.

If you can get LLVM IR for the cases where CGP widening causes issues, can you please attach them here so I can add them with a future patch ?

**Comment by jayfoad** [comment]
Here's one obvious;y artificial case:
```
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
; RUN: llc -march=amdgcn -mcpu=gfx1030 -amdgpu-codegenprepare-widen-16-bit-ops=0 < %s | FileCheck %s -check-prefix=NOPROMOTE
; RUN: llc -march=amdgcn -mcpu=gfx1030 -amdgpu-codegenprepare-widen-16-bit-ops=1 < %s | FileCheck %s -check-prefix=PROMOTE

define i16 @square(i16 inreg %v0) {
; NOPROMOTE-LABEL: square:
; NOPROMOTE:       ; %bb.0:
; NOPROMOTE-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
; NOPROMOTE-NEXT:    s_mul_i32 s4, s16, s16
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; NOPROMOTE-NEXT:    v_mov_b32_e32 v0, s4
; NOPROMOTE-NEXT:    s_setpc_b64 s[30:31]
;
; PROMOTE-LABEL: square:
; PROMOTE:       ; %bb.0:
; PROMOTE-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
; PROMOTE-NEXT:    s_and_b32 s4, 0xffff, s16
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_and_b32 s4, s4, 0xffff
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_and_b32 s4, s4, 0xffff
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_and_b32 s4, s4, 0xffff
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    s_mul_i32 s4, s4, s4
; PROMOTE-NEXT:    v_mov_b32_e32 v0, s4
; PROMOTE-NEXT:    s_setpc_b64 s[30:31]
  %v1 = mul i16 %v0, %v0
  %v2 = mul i16 %v1, %v1
  %v3 = mul i16 %v2, %v2
  %v4 = mul i16 %v3, %v3
  %v5 = mul i16 %v4, %v4
  %v6 = mul i16 %v5, %v5
  %v7 = mul i16 %v6, %v6
  %v8 = mul i16 %v7, %v7
  %v9 = mul i16 %v8, %v8
  ret i16 %v9
}
```

**Comment by jayfoad** [comment]
> I got stuck trying to implement the promotion logic in GlobalISel. I think it has to be done in the legalizer, but the legalizer needs UniformityInfo to do it. Whenever I tried doing it somewhere else I ran into issues, either combine loops or generating illegal instructions too late.

Generally legalizations that have to be different for SALU and VALU are implemented in RegBankSelect. For example see AND/OR/XOR, where SALU has 64-bit ops but VALU does not:

https://github.com/llvm/llvm-project/blob/a98707e285e08bbf20785bfe54190feab7eb3c91/llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp#L2414

**Comment by Pierre-vh** [comment]
The stack of patch ending with #131312 should help with regressions. We could almost do the switch now but there's a few regressions left with s/u min/max in DAGISel.

**Comment by Pierre-vh** [comment]
I think this is getting close to done. The last regressions I see are related to min/max in the DAG path (one or two more instructions everytime), and #140040

In a lot of cases we actually get much better codegen with i16 promotion in CGP disabled.
